lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,234 @@
+# type ('String' in the example):
+#
+#     lookup('the_name', String)
+#
+# When called with four arguments, **the name**, the **expected type**, the **merge** strategy, and a
+# **default value**, it returns the bound value with the given name, or the default after having asserted the value
+# has the given data type:
+#
+#     lookup('the_name', String, undef, 'Fred')
+#     lookup('the_name', Array[String], 'unique', [Fred])
+#
+# Using a lambda to provide a default value by calling a function:
+#
+#     lookup('the_size', Integer[1,100]) |$name| {
+#       obtain_size_default()
+#     }
+#
+# There are two ways to make lookup return undef when no matching key was found instead of raising an error.
+# Either call it with four arguments (the `merge` argument must be present even when using the default strategy
+# to ensure that the four argument variant is used):
+#
+#      $are_you_there = lookup('peekaboo', Optional[String], undef, undef)
+#
+# Or call it with a block that produces an undef value:
+#
+#      $are_you_there = lookup('peekaboo', Optional[String]) |$name| { undef }
+#
+# - Since 4.0.0
+Puppet::Functions.create_function(:lookup, Puppet::Functions::InternalFunction) do
+  name_t = 'Variant[String,Array[String]]'
+  value_type_t = 'Optional[Type]'
+  default_value_t = 'Any'
+  override_t = 'Optional[Hash[String,Any]]'
+  default_values_hash_t = 'Optional[Hash[String,Any]]'
+  merge_t = 'Optional[Variant[String[1],Hash[String,Scalar]]]'
+  block_t = "Callable[#{name_t}]"
+  option_pairs =
+      "value_type=>#{value_type_t},"\
+      "default_value=>Optional[#{default_value_t}],"\
+      "override=>#{override_t},"\
+      "default_values_hash=>#{default_values_hash_t},"\
+      "merge=>#{merge_t}"\
+
+  dispatch :lookup_1 do
+    scope_param
+    param name_t, :name
+    param value_type_t, :value_type
+    param merge_t, :merge
+    arg_count(1, 3)
+  end
+
+  dispatch :lookup_2 do
+    scope_param
+    param name_t, :name
+    param value_type_t, :value_type
+    param merge_t, :merge
+    param default_value_t, :default_value
+  end
+
+  dispatch :lookup_3 do
+    scope_param
+    param name_t, :name
+    param value_type_t, :value_type
+    param merge_t, :merge
+    arg_count(1, 3)
+    required_block_param block_t, :block
+  end
+
+  # Lookup without name. Name then becomes a required entry in the options hash
+  dispatch :lookup_4 do
+    scope_param
+    param "Struct[{name=>#{name_t},#{option_pairs}}]", :options_hash
+    optional_block_param block_t, :block
+  end
+
+  # Lookup using name and options hash.
+  dispatch :lookup_5 do
+    scope_param
+    param 'Variant[String,Array[String]]', :name
+    param "Struct[{#{option_pairs}}]", :options_hash
+    optional_block_param block_t, :block
+  end
+
+  def lookup_1(scope, name, value_type=nil, merge=nil)
+    Puppet::Pops::Binder::Lookup.lookup(scope, name, value_type, nil, false, {}, {}, merge)
```

Comment:
We should refactor the Lookup helper from Puppet::Pops::Binder::Lookup to Puppet::Pops::Lookup since it is not binder specific after the changes in this PR.